packages/@sanity/form-builder/src/inputs/ObjectInput/ObjectFieldSet.tsx

```diff
@@ -0,0 +1,97 @@
+  markers: Marker[]
+}
+
+/**
+ * A specialized component for object "fieldsets", e.g. a group of object fields as defined by the fieldsets option
+ * on the object type https://www.sanity.io/docs/object-type#fieldsets-b8b5507db1d3
+ */
+export const ObjectFieldSet = forwardRef(function ObjectFieldSet(
+  props: Props,
+  forwardedRef: ForwardedRef<HTMLElement>
+) {
+  const {fieldset, focusPath, children, level, presence, markers, onFocus, ...rest} = props
+  const columns = fieldset.options && fieldset.options.columns
+
+  const collapsibleOpts = getCollapsedWithDefaults(fieldset.options, level)
+
+  const fieldNames = useMemo(() => fieldset.fields.map((f) => f.name), [fieldset.fields])
+
+  const [isCollapsed, setCollapsed] = React.useState(collapsibleOpts.collapsed)
+
+  const childPresence = useMemo(() => {
+    return isCollapsed && presence.length > 0
+      ? presence.filter(
+          (item) => typeof item.path[0] === 'string' && fieldNames.includes(item.path[0])
+        )
+      : EMPTY_ARRAY
+  }, [fieldNames, isCollapsed, presence])
+
+  const childMarkers = useMemo(() => {
+    return markers.length === 0
+      ? markers
+      : markers.filter(
+          (item) => typeof item.path[0] === 'string' && fieldNames.includes(item.path[0])
+        )
+  }, [fieldNames, markers])
+
+  const handleToggleFieldset = React.useCallback(
+    (nextCollapsed) => {
+      if (nextCollapsed) {
+        // note: technically, when collapsing this fieldset we remove focus from whatever field (if any) inside
+        // it that currently has focus, so it might be tempting to emit a new focus path for the "parent object",
+        // e.g. onFocus([]), but this can in some cases have the unintended consequence of creating page jumps, since it
+        // will put focus in the the first field of the parent object input, which may be rendered far away from the current fieldset.
+        // For this reason, keep the focus path where it is, and just "locally" collapse this fieldset
+        setCollapsed(true)
+      } else {
+        onFocus([fieldNames[0]])
+        setCollapsed(false)
+      }
+    },
+    [onFocus, fieldNames]
+  )
+
+  React.useEffect(() => {
+    const hasFocusWithin =
+      focusPath.length > 0 && fieldNames.some((fieldName) => focusPath[0] === fieldName)
+
+    if (hasFocusWithin) {
+      setCollapsed(false)
+    }
+  }, [fieldNames, focusPath])
+
+  return (
+    <FormFieldSet
+      {...rest}
+      key={fieldset.name}
+      title={fieldset.title}
+      description={fieldset.description}
+      level={level + 1}
+      columns={columns}
+      collapsible={collapsibleOpts.collapsible}
+      collapsed={isCollapsed}
+      onToggle={handleToggleFieldset}
+      __unstable_presence={isCollapsed ? childPresence : EMPTY_ARRAY}
+      __unstable_changeIndicator={false}
+      __unstable_markers={childMarkers}
+      ref={isCollapsed ? forwardedRef : null}
+    >
+      {isCollapsed ? null : children()}
```

Comment:
Can the `children` callback be passed a function instead, since `FormFieldSet` now can handle that?

```tsx
<FormFieldSet {...}>
  {isCollapsed ? null : children}
</FormFieldSet>
```

Or maybe even just:

```tsx
<FormFieldSet {...}>
  {children}
</FormFieldSet>
```

Reply:
Oh yes, great catch!

Reply:
Fixed in https://github.com/sanity-io/sanity/pull/2482/commits/58bc882be9eebf19f22d8260c3d8b3322e9053ef